### ISSUE TITLE
Subprocess support for scripting

### ISSUE THREAD

**Issue by mollymorphic** [zetavm/zetavm#94]
In order to be useful for minimal shell/scripting-type tasks zeta should provide the ability to create & control subprocesses.

A reasonable first implementation would:

- have a synchronous API
- be implemented with or based on the `popen` function: [http://man7.org/linux/man-pages/man3/popen.3.html](http://man7.org/linux/man-pages/man3/popen.3.html)
- have at least one function which accepts a command to run with stdin and/or stdout being relegated to the parent (zeta) process
- have at least one function which accepts a command to run with stdin and/or stdout being provided as arguments (pointers to strings, buffers, etc)

Note that this API has a lot of potential for yak-shaving as the low level will be pretty awkward to use. The bare minimum to allow some scripting should be completed first, later when the standard lib is more fleshed out with allocation/concurrency/string functions more convenient wrappers can be created.

Higher level functions may include:

- a function which accepts a command to run and data to pipe into it's stdin
- a unction which accepts a command to run and a callback function which will receive the output of the command
- etc

**Comment by maximecb** [comment]
Yeah, what scares me a little bit here is that I've worked with Python's popen, and it was a mess. There were a lot of corner cases to deal with. For instance, it seems that some programs can tell if you've read their stdout, and will block until you've done so. Then, of course, there are platform-specific differences.

Another issue is that, from what I'm reading, `popen` allows you to read from the subprocess's `stdout`, or write to its `stdin`, but not both. I guess if I have to choose, I would prefer to get the stdout. I don't think we actually need a callback. We can return an object with a field for the return value, and one for the output as a string.

**Comment by mollymorphic** [comment]
Ya, you can only get one or the other as a pipe from `popen()` I would agree if you had to choose one then getting stdout is more generally useful - but if possible it'd be nice to expose both.

In either case whatever other streams aren't captured are just delegated to the parent process which helps some - you can still get to them, it's just awkward. More complicated flows might be built with `fork()` or by creating `pipe()`s and getting the subprocess to use them for stdin/stdout.

It will probably take quite some time for this library to be built up to something smooth to use and that covers all the corner cases, but the basic functions will at least enable simple use cases.

**Comment by maximecb** [comment]
Possibly, we could create a `core/system` library, and have an `exec` command which works as follows:

```
system = import "core/system/0";
output = system.exec(commandString, inputString);
```

The `inputString` would get written to the subprocess's `stdin` after it's started. Then we could read the subprocess's stdout and stderr and put those into strings.

```
output.stdout // read from stdout
output.stderr // read from stderr
output.ret // return value after termination
```

Another option would be to combine both stdout and stderr into a single "output" string.

**Comment by boomshroom** [comment]
A mechanism for starting processes would be nice. Once that's done, one direction to continue could be adding some way of providing source code to zeta through stdin with some way to select the language package. maybe something like `zeta -i lang/plush/0` and it would read plush code interactively from stdin as a repl or shell.

This by no means has to be implemented and would probably require a separate function to parse individual expressions rather than whole packages, but it seems to be a logical next step and would enable the use of zeta and its language packages as a command line shells akin to bash. With this sort of thing, I enjoy looking at the roadmap and thinking about what could be beyond.

**Comment by maximecb** [comment]
@boomshroom There may be a case for passing code through stdin. Using Zeta as a bash replacement seems beyond the scope of this project though. Seems we would need to expose a ton of posix functionality to make that work.

**Comment by boomshroom** [comment]
It's more of a stretch goal. I would not want time taken away from the core to implement it. It's just a possible direction when everything is more or less complete.

**Comment by maximecb** [comment]
Will close this for now since it isn't needed at the moment and we don't have a very clear plan.
